### ISSUE TITLE
Consider throwing if AdvanceTo is called after writer is complete and reader hasn't consumed enough

### ISSUE THREAD

**Issue by davidfowl** [dotnet/corefx#30783]
When using pipelines, if the writer is done writing (therefore ReadResult.IsCompleted is true) and `AdvanceTo` doesn't consume everything then we should throw because the reader cannot make any progress. 

cc: @pakrym

**Comment by mgravell** [comment]
> doesn't consume everything

Is it more correct to throw if it doesn't consume *anything*? meaning: consuming *everything* requires that reader code deals with as many frames as possible in each call; that's *desirable*, but is it *enforced*? by which I mean: is this a valid use-case:

```c#
while (true)
{
    var read = await reader.ReadAsync();
    // ...
    var buffer = read.Buffer;
    TryProcessSingleFrame(ref buffer); // slices buffer if successful
    reader.AdvanceTo(buffer.Start, buffer.End);
}
```

Now imagine that we have 3 and a half frames in the queue when the writer completes. There are 3 *perfectly good frames* there; IMO we shouldn't bomb at the end of the first one just because they didn't read all 3. I'm perfectly happy for it to fail on the 4th attempt, when they don't change `buffer`.

Hence again: Is it more correct to throw if it doesn't consume *anything*, rather than if it doesn't consume *everything*?

**Comment by davidfowl** [comment]
Yes I think that's more correct. If forward progress hasn't been made then we throw.

**Comment by pakrym** [comment]
But only in the cases where everything is observed, right?

**Comment by mgravell** [comment]
@pakrym if they're not making progress... does it matter? there *isn't going to be anything else*; maybe "progress" means "consuming at least one byte *or* advancing 'examined' by at least one byte" (does it actually store `examined`?) - but my fear is that there could be code that misreports `examined`; misreporting `consumed` is much harder to screw up.

**Comment by pakrym** [comment]
I don't want to disable `TryRead*(IPipeReader)` pattern by being too strict with exceptions.

**Comment by davidfowl** [comment]
Yes, we can say if the entire buffer is examined and nothing was consumed then throw.

**Comment by davidfowl** [comment]
Current proposal, we have a setting that determines when to throw on not making progress.

**Comment by davidfowl** [comment]
@pakrym took a quick look and this adds a bunch of complexity and state to pipes, we think we want to avoid this in 3.0 so I'm moving this to future
